File: 114_FlattenBinaryTreeToLinkedList.cpp
// Topic: Flatten Binary Tree to Linked List
// Description: Flatten tree to linked list in-place following preorder (right pointers as next).
// Approach: Use iterative stack or reverse preorder recursion. We'll use iterative stack (O(n)
// time, O(n) stack).

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* buildTreeLevel(const vector<int>& vals)
{
    if (vals.empty())
        return nullptr;
    int n = vals.size();
    if (vals[0] == INT_MIN)
        return nullptr;
    vector<TreeNode*> nodes(n, nullptr);
    for (int i = 0; i < n; ++i)
        if (vals[i] != INT_MIN)
            nodes[i] = new TreeNode(vals[i]);
    for (int i = 0; i < n; ++i)
        if (nodes[i])
        {
            int l = 2 * i + 1, r = 2 * i + 2;
            if (l < n)
                nodes[i]->left = nodes[l];
            if (r < n)
                nodes[i]->right = nodes[r];
        }
    return nodes[0];
}

void flatten(TreeNode* root)
{
    if (!root)
        return;
    stack<TreeNode*> st;
    st.push(root);
    TreeNode* prev = nullptr;
    while (!st.empty())
    {
        TreeNode* cur = st.top();
        st.pop();
        if (prev)
        {
            prev->left = nullptr;
            prev->right = cur;
        }
        if (cur->right)
            st.push(cur->right);
        if (cur->left)
            st.push(cur->left);
        prev = cur;
    }
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<vector<int>> tests = {{1, 2, 5, 3, 4, INT_MIN, 6}, {}};

    for (int t = 0; t < (int)tests.size(); ++t)
    {
        auto vals = tests[t];
        TreeNode* root = buildTreeLevel(vals);
        flatten(root);
        cout << "Test " << (t + 1) << ": ";
        TreeNode* cur = root;
        if (!cur)
            cout << "(empty)";
        while (cur)
        {
            cout << cur->val;
            if (cur->right)
                cout << ' ';
            cur = cur->right;
        }
        cout << '\n';
    }
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(n) stack
*/

File: 114_FlattenBinaryTreeToLinkedList.py
# Topic: Flatten Binary Tree to Linked List
# Description: Flatten tree to linked list in-place following preorder (right pointers as next).
# Approach: Iterative stack or reverse preorder recursion. Using iterative stack.

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree_level(vals: List[int]) -> Optional[TreeNode]:
    if not vals or vals[0] is None:
        return None
    nodes = [TreeNode(v) if v is not None else None for v in vals]
    for i, node in enumerate(nodes):
        if node:
            l, r = 2 * i + 1, 2 * i + 2
            if l < len(nodes):
                node.left = nodes[l]
            if r < len(nodes):
                node.right = nodes[r]
    return nodes[0]


class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        if not root:
            return
        stack = [root]
        prev = None
        while stack:
            cur = stack.pop()
            if prev:
                prev.left = None
                prev.right = cur
            if cur.right:
                stack.append(cur.right)
            if cur.left:
                stack.append(cur.left)
            prev = cur


def print_flattened(root: Optional[TreeNode]) -> List[int]:
    result = []
    while root:
        result.append(root.val)
        root = root.right
    return result


if __name__ == "__main__":
    tests = [[1, 2, 5, 3, 4, None, 6], []]
    for i, vals in enumerate(tests):
        root = build_tree_level(vals) if vals else None
        Solution().flatten(root)
        print(f"Test {i + 1}: {print_flattened(root)}")

"""
Time Complexity: O(n)
Space Complexity: O(n) stack
"""
